src/tdigest_api.c

```diff
@@ -0,0 +1,693 @@
+		ts_tdigest_trans_compress(incoming_tts);
+
+		receiver_tts = ts_tdigest_startup(0);
+		receiver_tts->td = ts_tdigest_copy(incoming_tts->td);
+	}
+	else if (PG_ARGISNULL(1)) // if nothing to merge in, merged value is current value
+	{
+		receiver_tts = (TDigestTState *) PG_GETARG_VARLENA_P(0);
+	}
+	else // merge states if both are non-NULL
+	{
+		receiver_tts = (TDigestTState *) PG_GETARG_VARLENA_P(0);
+		incoming_tts = (TDigestTState *) PG_GETARG_VARLENA_P(1);
+
+		ts_tdigest_trans_merge(receiver_tts, incoming_tts);
+	}
+
+	MemoryContextSwitchTo(old);
+
+	PG_RETURN_POINTER(receiver_tts);
+}
+
+/*
+ * Final function for aggregate
+ * Returns: TDigest
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_final_sql);
+Datum
+ts_tdigest_final_sql(PG_FUNCTION_ARGS)
+{
+	TDigestTState *tts;
+
+	if (!AggCheckCallContext(fcinfo, NULL))
+	{
+		/* cannot be called directly because of internal-type argument */
+		elog(ERROR, "ts_tdigest_final_sql called in non-aggregate context");
+	}
+
+	/* use getarg_pointer because the transition state is a pointer, only
+	 * the internal TDigest is TOAST'ed */
+	tts = PG_ARGISNULL(0) ? NULL : (TDigestTState *) PG_GETARG_POINTER(0);
+
+	if (tts == NULL || tts->td == NULL)
+		PG_RETURN_NULL();
+
+	/* compress before returning to ensure all centroids merged into tdigest */
+	ts_tdigest_trans_compress(tts);
+
+	PG_RETURN_POINTER(tts->td);
+}
+
+/********************
+ * CLIENT FUNCTIONS *
+ ********************/
+
+/*
+ * Returns the approximate value for a percentile using a TDigest
+ * Args: TDigest structure, Percentile
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_percentile_sql);
+Datum
+ts_tdigest_percentile_sql(PG_FUNCTION_ARGS)
+{
+	TDigest *t;
+	float8 q;
+
+	if (PG_ARGISNULL(0))
+		PG_RETURN_NULL();
+
+	t = (TDigest *) PG_GETARG_VARLENA_P(0);
+	q = PG_GETARG_FLOAT8(1);
+
+	PG_RETURN_FLOAT8(ts_tdigest_percentile(t, q));
+}
+
+/*
+ * Returns the approximate value of the CDF using a TDigest
+ * Args: TDigest stucture, CDF value
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_cdf_sql);
+Datum
+ts_tdigest_cdf_sql(PG_FUNCTION_ARGS)
+{
+	TDigest *t;
+	float8 x;
+
+	if (PG_ARGISNULL(0))
+		PG_RETURN_NULL();
+
+	t = (TDigest *) PG_GETARG_VARLENA_P(0);
+	x = PG_GETARG_FLOAT8(1);
+
+	PG_RETURN_FLOAT8(ts_tdigest_cdf(t, x));
+}
+
+/*
+ * Returns the total number of data points put into this TDigest thus far
+ * Avoids user needing to call count() on a dataset they've already put into a TDigest
+ */
+TS_FUNCTION_INFO_V1(ts_tdigest_count_sql);
+Datum
+ts_tdigest_count_sql(PG_FUNCTION_ARGS)
+{
+	TDigest *t;
+
+	if (PG_ARGISNULL(0))
+		PG_RETURN_INT32(0);
+
+	t = (TDigest *) PG_GETARG_VARLENA_P(0);
+
+	PG_RETURN_INT32(ts_tdigest_count(t));
+}
+
+/***************
+ * TDIGEST I/O *
+ ***************/
+
+/*
+ * Converts a tdigest to a portable binary representation
+ */
+static bytea *
+tdigest_send(StringInfo buf, TDigest *t)
+{
+	int i;
+
+	pq_begintypsend(buf);
+
+	/* send the tdigest size */
+	pq_sendint64(buf, ts_tdigest_size(t));
```

Comment:
see my comment about size of tdigest (tdigest.c). once we resolve that, this should either read directly from VARSIZE(t) or use pq_sendint32.